cpp/include/raft/distance/kernels/kernel_matrices.cuh

```diff
@@ -0,0 +1,741 @@
+/*
+ * Copyright (c) 2019-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "gram_matrix.cuh"
+#include <raft/util/cuda_utils.cuh>
+
+#include <raft/distance/distance.cuh>
+#include <raft/linalg/gemm.cuh>
+#include <raft/sparse/linalg/norm.cuh>
+
+namespace raft::distance::kernels {
+
+/** Epiloge function for polynomial kernel without padding.
+ * Calculates output = (gain*in + offset)^exponent
+ * @param inout device vector in column major format, size [len]
+ * @param len array length
+ * @param exponent
+ * @param gain
+ * @param offset
+ */
+template <typename math_t, typename exp_t>
+__global__ void polynomial_kernel_nopad(
```

Comment:
The reason these objects were in detail originally and only the public bits exposed via typedef (`using`) was because these kernels shouldn't be exposed to the public APIs. Also, this class-based (mostly-stateless) design for host-callable computation runs counter to all of the other APIs in RAFT. I think we've all agreed that we should flatten these APIs into a series of stateless functions that can be called and have their state passed into them, but we haven't yet prioritized that, thus I wanted to keep the code out of the docs and obscure it from being discoverable by users until we have the new "formal" API exposed. I suggest we move these back into the `detail` namespace and expose them with `using` into the public namespace but also expose the other bits we need from this API to cuml in the same way (with `using`) so that cuml isn't calling APIs in the detail namespace. 

I do apologize if my wording has been confusing in previous reviews.